src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1359,17 +1290,26 @@ InstanceKlass* SystemDictionary::load_instance_class_impl(Symbol* class_name, Ha
 
     InstanceKlass* spec_klass = vmClasses::ClassLoader_klass();
 
-    // Call public unsynchronized loadClass(String) directly for all class loaders.
-    // For parallelCapable class loaders, JDK >=7, loadClass(String, boolean) will
-    // acquire a class-name based lock rather than the class loader object lock.
-    // JDK < 7 already acquire the class loader lock in loadClass(String, boolean).
-    JavaCalls::call_virtual(&result,
-                            class_loader,
-                            spec_klass,
-                            vmSymbols::loadClass_name(),
-                            vmSymbols::string_class_signature(),
-                            string,
-                            CHECK_NULL);
+    if (is_parallelCapable(class_loader)) {
```

Comment:
You could move the dispatching logic into Java as well .

Reply:
I had it there but Mandy thought the dispatching code should be in the JVM.  Moving the dispatch there allows me to add a flag to control this.